File: packages/protocol/src/primitives.ts
import { z } from 'zod'

/**
 * 协议共用的标量与小结构。
 *
 * 这里的 `usage` / `annotation` / `reasoningKind` 等类型在旧实现里是直接把 Koaks 框架的
 * `Usage` / `Annotation` / `ModelEvent` 透传到 renderer 的（`timeline-model.ts` 曾直接
 * `import type { Annotation, ModelEvent } from '@koaks/node'`）。KAP 明确禁止框架类型外泄：
 * 协议持有自己的等价定义，Koaks 只是 server 的一个内部实现细节。
 */

export const epochMillisSchema = z.number().int().nonnegative()

export const idSchema = z.string().min(1)

/** 单调递增的事件序号，客户端用它做断线补发的游标。 */
export const sequenceSchema = z.number().int().nonnegative()

export const usageSchema = z.object({
  promptTokens: z.number().int().nonnegative(),
  completionTokens: z.number().int().nonnegative(),
  totalTokens: z.number().int().nonnegative(),
  cachedInputTokens: z.number().int().nonnegative(),
  reasoningOutputTokens: z.number().int().nonnegative()
})
export type UsageDto = z.infer<typeof usageSchema>

export const ZERO_USAGE: UsageDto = {
  promptTokens: 0,
  completionTokens: 0,
  totalTokens: 0,
  cachedInputTokens: 0,
  reasoningOutputTokens: 0
}

/**
 * `legacy` 表示模型没有区分推理内容的种类。保留它是因为部分供应商只回一段未标注的推理文本，
 * 客户端需要能把它和 `summary` / `raw` 区分开来展示。
 */
export const reasoningKindSchema = z.enum(['summary', 'raw', 'legacy'])
export type ReasoningKind = z.infer<typeof reasoningKindSchema>

/** 结构化输出的收尾步骤与普通步骤要区分，插件的 before-model hook 通常应跳过收尾步骤。 */
export const modelCallPhaseSchema = z.enum(['normal', 'structured_finalization'])
export type ModelCallPhase = z.infer<typeof modelCallPhaseSchema>

export const annotationSchema = z.discriminatedUnion('type', [
  z.object({
    type: z.literal('url_citation'),
    url: z.string(),
    title: z.string().optional(),
    startIndex: z.number().int().nonnegative().optional(),
    endIndex: z.number().int().nonnegative().optional()
  }),
  z.object({
    type: z.literal('file_citation'),
    fileId: z.string(),
    filename: z.string().optional(),
    startIndex: z.number().int().nonnegative().optional(),
    endIndex: z.number().int().nonnegative().optional()
  }),
  z.object({
    type: z.literal('generic'),
    kind: z.string(),
    payload: z.string()
  })
])
export type AnnotationDto = z.infer<typeof annotationSchema>

/** 模型请求里引用的条目锚点，客户端用它把增量事件归并到同一段内容上。 */
export const itemRefSchema = z.string().min(1)

/**
 * 只接受 http/https 的 URL。
 *
 * 比 `z.string().url()` 更严：后者会放过 `ftp://` 之类。限定 scheme 的另一个原因是 Kotlin 侧
 * 的 `require()` 校验必须能给出完全一致的判定，否则 conformance fixture 会在两侧结论不同。
 */
export const httpUrlSchema = z
  .string()
  .trim()
  .url()
  .refine((value) => /^https?:\/\//i.test(value), {
    message: 'URL 必须以 http:// 或 https:// 开头'
  })

File: agent/protocol/src/commonMain/kotlin/dev/kowork/protocol/Primitives.kt
package dev.kowork.protocol

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonClassDiscriminator

/**
 * 协议共用的标量与小结构。
 *
 * 这里的 [Usage] / [KapAnnotation] / [ReasoningKind] 都是**协议自己的**类型，不是 Koaks 框架
 * 类型的别名。旧实现把 Koaks 的 `Usage` / `Annotation` / `ModelEvent` 一路透传到 Electron
 * renderer，客户端因此直接 import 了框架类型。KAP 明确切断这条链路：Koaks 只是 server 内部的
 * 实现细节，server 负责把框架事件映射成协议事件。
 */

@Serializable
data class Usage(
    val promptTokens: Long,
    val completionTokens: Long,
    val totalTokens: Long,
    val cachedInputTokens: Long,
    val reasoningOutputTokens: Long
) {
    init {
        requireNonNegative(promptTokens, "usage.promptTokens")
        requireNonNegative(completionTokens, "usage.completionTokens")
        requireNonNegative(totalTokens, "usage.totalTokens")
        requireNonNegative(cachedInputTokens, "usage.cachedInputTokens")
        requireNonNegative(reasoningOutputTokens, "usage.reasoningOutputTokens")
    }

    companion object {
        val ZERO = Usage(0, 0, 0, 0, 0)
    }
}

/**
 * 推理内容的种类。
 *
 * [LEGACY] 表示供应商没有区分推理种类 —— 部分供应商只回一段未标注的推理文本，客户端需要
 * 能把它和 [SUMMARY] / [RAW] 区分开来展示。
 */
@Serializable
enum class ReasoningKind {
    @SerialName("summary")
    SUMMARY,

    @SerialName("raw")
    RAW,

    @SerialName("legacy")
    LEGACY
}

/** 结构化输出的收尾步骤与普通步骤要区分，插件的 before-model hook 通常应跳过收尾步骤。 */
@Serializable
enum class ModelCallPhase {
    @SerialName("normal")
    NORMAL,

    @SerialName("structured_finalization")
    STRUCTURED_FINALIZATION
}

@OptIn(ExperimentalSerializationApi::class)
@Serializable
@JsonClassDiscriminator("type")
sealed interface KapAnnotation {
    @Serializable
    @SerialName("url_citation")
    data class UrlCitation(
        val url: String,
        val title: String? = null,
        val startIndex: Long? = null,
        val endIndex: Long? = null
    ) : KapAnnotation {
        init {
            startIndex?.let { requireNonNegative(it, "annotation.startIndex") }
            endIndex?.let { requireNonNegative(it, "annotation.endIndex") }
        }
    }

    @Serializable
    @SerialName("file_citation")
    data class FileCitation(
        val fileId: String,
        val filename: String? = null,
        val startIndex: Long? = null,
        val endIndex: Long? = null
    ) : KapAnnotation {
        init {
            startIndex?.let { requireNonNegative(it, "annotation.startIndex") }
            endIndex?.let { requireNonNegative(it, "annotation.endIndex") }
        }
    }

    @Serializable
    @SerialName("generic")
    data class Generic(
        val kind: String,
        val payload: String
    ) : KapAnnotation
}
